src/main/core/worker.rs

```diff
@@ -285,16 +288,51 @@ impl Worker {
     }
 
     pub fn increment_object_alloc_counter(s: &str) {
-        let s = std::ffi::CString::new(s).unwrap();
-        unsafe { cshadow::worker_increment_object_alloc_counter(s.as_ptr()) }
+        if !USE_OBJECT_COUNTERS.load(std::sync::atomic::Ordering::Relaxed) {
+            return;
+        }
+
+        Worker::with_mut(|w| {
+            w.object_alloc_counter.add_one(s);
+        })
+        .unwrap_or_else(|| {
+            // no live worker; fall back to the shared manager counter
+            let s = std::ffi::CString::new(s).unwrap();
+            unsafe { cshadow::manager_increment_object_alloc_counter_global(s.as_ptr()) };
+        });
     }
 
     pub fn increment_object_dealloc_counter(s: &str) {
-        let s = std::ffi::CString::new(s).unwrap();
-        unsafe { cshadow::worker_increment_object_dealloc_counter(s.as_ptr()) }
+        if !USE_OBJECT_COUNTERS.load(std::sync::atomic::Ordering::Relaxed) {
+            return;
+        }
+
+        Worker::with_mut(|w| {
+            w.object_dealloc_counter.add_one(s);
+        })
+        .unwrap_or_else(|| {
+            // no live worker; fall back to the shared manager counter
+            let s = std::ffi::CString::new(s).unwrap();
+            unsafe { cshadow::manager_increment_object_dealloc_counter_global(s.as_ptr()) };
+        });
+    }
+
+    pub fn worker_add_syscall_counts(syscall_counts: &Counter) {
+        Worker::with_mut(|w| {
+            w.syscall_counter.add_counter(syscall_counts);
+        })
+        .unwrap_or_else(|| {
+            // no live worker; fall back to the shared manager counter
+            unsafe { cshadow::manager_add_syscall_counts_global(syscall_counts) };
+        });
     }
 }
 
+/// Enable object counters. Should be called near the beginning of the program.
+pub fn enable_object_counters() {
+    USE_OBJECT_COUNTERS.store(true, std::sync::atomic::Ordering::SeqCst);
```

Comment:
Optional nit - I think `Relaxed` would be fine here...? This atomic isn't protecting any other variables.

Reply:
Changed.